### PR TITLE
feat(advanced-options): add pure component for Scaling

### PR DESCRIPTION
**Pull request by debsmita1**
This PR includes:

- A  pure component for Advanced Options: Scaling
- The Number Spinner for field `Replicas` displays an error when non-integer or negative integer is entered

Scaling

Refer Jira Story : https://jira.coreos.com/browse/ODC-821

### CODE REVIEW

**Comment by rohitkrai03** [comment]
@debsmita1 Don't integrate into git import form right now, that whole file is going to go away or re-written from scratch. The story is just about creating the component.

**Review by sahil143 — COMMENTED** [review]
@debsmita1 Scaling shouldn't allow values below zero if they are not valid. Disable the `-` button when value hits zero or accepts the minimum valid value as a prop.
Peek 2019-05-31 00-15

File: frontend/public/extend/devconsole/components/ImportFlowForm/ImportFlowForm.tsx [hunk 1]
```diff
@@ -51,6 +52,7 @@ export interface State {
     [key: string]: string;
   };
   showSourceSecretDropDown: boolean;
+  replicas: string;
```
Comment (abhinandan13jan) [inline frontend/public/extend/devconsole/components/ImportFlowForm/ImportFlowForm.tsx]: Is there anything specific that this is typed as a 'string' and not as a 'number'?

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    ) {
+      return false;
+    }
+    return _.isInteger(_.toNumber(this.props.replicas));
+  };
+
+  render() {
+    return (
+      <React.Fragment>
+        <h2 className="odc-scaling">
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: We need common components for these form headers. Is there not already one available?

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+          Scaling
+          <FieldLevelHelp content="Scaling defines the number of running instances of your built image." />
+        </h2>
+        <h4 className="odc-scaling-info">Replicas are scaled manually based on CPU usage. </h4>
+        <FormGroup className={this.isInteger() ? '' : 'has-error'}>
+          <ControlLabel>Replicas</ControlLabel>
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Does this label get linked to the form field? ie. when you click it does it focus on the number input?

File: frontend/public/extend/devconsole/components/ImportFlowForm/ImportFlowForm.tsx (hunk 1 above)
Comment (debsmita1) [reply frontend/public/extend/devconsole/components/ImportFlowForm/ImportFlowForm.tsx]: Changed it to `number`

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  };
+
+  isPositiveInteger = () => {
+    return _.isInteger(this.props.replicas) && this.props.replicas >= 0;
+  };
+
+  render() {
+    return (
+      <React.Fragment>
+        <h2 className="odc-scaling">Scaling</h2>
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Need common components for headings so we do not all create custom CSS.

Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Fixed https://github.com/talamer/console/pull/118/commits/4b235476926e963291946907f926f3fe335f40dc

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx (hunk 3 above)
Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Fixed it by adding `controlId` in the `FormGroup` https://github.com/talamer/console/pull/118/commits/4b235476926e963291946907f926f3fe335f40dc

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx (hunk 2 above)
Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Used `Title` patternfly component https://github.com/talamer/console/pull/118/commits/4b235476926e963291946907f926f3fe335f40dc

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+  isPositiveInteger = () => {
+    return _.isInteger(this.props.replicas) && this.props.replicas >= 0;
+  };
+
+  render() {
+    return (
+      <React.Fragment>
+        <Title size="lg">Scaling</Title>
+        <Text className="odc-scaling-info" component={TextVariants.h5}>Replicas are scaled manually based on CPU usage.</Text>
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Why is this anything other than normal text?
Can we remove this `odc-scaling-info`?
@rohitkrai03 please work with the rest of the team making forms to align on a set of common patterns to adhere to. I don't see why any one form component would need custom styles.

Comment (rohitkrai03) [reply frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: @debsmita1 Why are you using some components from `Patternfly 4` and some from `Patternfly 3`? I think it should be a consistent experience and right now we are using patternfly 3 for forms so it makes sense to use it everywhere in the forms.

**Comment by christianvogt** [comment]
By validating on the fly in the render function, the parent form will never know of the validation state.
We should create utilities for each form component to support validating form fields. The parent form can use these utilities to manage form validation. This would mean that the field errors need to be given as props.
@rohitkrai03 please align with the team on the approach for form components so that we are all doing the same thing and it fits with what you propose with formik

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

**Comment by rohitkrai03** [comment]
@christianvogt Yes, i agree with the validation method. The forms can just provide a utility function to validate the fields and then it can be used by parent form to validate the fields. Formik has both the options of validating whole form at once and field level validation. Some of the fields would need field level validation such as Git URl which needs async call for git source analysis and build tool detection.

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx (hunk 5 above)
Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Replaced the `Text` component with `div` https://github.com/talamer/console/commit/ad90aaa52292b1bfcc76a3d620460b238ee5b98a

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+export const isPositiveInteger = (replicas: number) => {
+  return _.isInteger(replicas) && replicas >= 0
+    ? ''
+    : 'Replicas must be an integer value greater than or equal to 0.';
+};
+
+const Scaling: React.FC<ScalingProps> = (props) => {
+  const onChangeData: React.ReactEventHandler<HTMLInputElement> = (event) => {
+    props.onChange(event.currentTarget.valueAsNumber, props.replicaError);
```
Comment (rohitkrai03) [inline frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Why are you sending props.replicaError with `onChange`?

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  };
+
+  const changeValueBy = (operation) => {
+    props.onChange(_.toInteger(props.replicas) + operation, props.replicaError);
```
Comment (rohitkrai03) [inline frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Same as above.

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+  onChange: (replicas: number, replicaError: string) => void;
+}
+
+export const isPositiveInteger = (replicas: number) => {
+  return _.isInteger(replicas) && replicas >= 0
+    ? ''
+    : 'Replicas must be an integer value greater than or equal to 0.';
+};
+
+const Scaling: React.FC<ScalingProps> = (props) => {
```
Comment (rohitkrai03) [inline frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Destructure the props here.

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+      <div className="co-section-heading-tertiary">
+        Replicas are scaled manually based on CPU usage.
+      </div>
+      <FormGroup controlId="scaling-replicas" className={props.replicaError ? 'has-error' : ''}>
+        <ControlLabel>Replicas</ControlLabel>
+        <NumberSpinner
+          id="scaling-replicas"
+          className="form-control"
+          value={props.replicas}
+          onChange={onChangeData}
```
Comment (rohitkrai03) [inline frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: `NumberSpinner` component doesn't accept or use `onChange`. So you can remove this.

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+          changeValueBy={changeValueBy}
```
Comment (rohitkrai03) [inline frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]:
```suggestion
          changeValueBy={(operation: number) => onChange(_.toInteger(replicas) + operation)}
```

**Review by rohitkrai03 — CHANGES_REQUESTED** [review]

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx (hunk 10 above)
Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Done https://github.com/talamer/console/commit/016aa32a785bc1e8032f0e610309e203573399d7

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx (hunk 8 above)
Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Done 016aa32

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx (hunk 6 above)
Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Removed it 016aa32

**Comment by debsmita1** [comment]
@rohitkrai03 The `onChange` prop in the `NumberSpinner` is used in cases where user enter values manually

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx [hunk 11]
```diff
@@ -0,0 +1,46 @@
+/* eslint-disable no-undef, dot-notation */
+import * as React from 'react';
+import * as _ from 'lodash-es';
+import { FormGroup, ControlLabel, HelpBlock } from 'patternfly-react';
+import { NumberSpinner } from '../../../../../components/utils';
+
+interface ScalingProps {
+  replicas: number;
+  replicaError: string;
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: `replicaError` and `replica` should be optional.

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+  onChange: (replicas: number) => void;
+}
+
+export const isPositiveInteger = (replicas: number) => {
```
Comment (christianvogt) [inline frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx:13]: This is fine for now to be here. We can align the validation strategy with other form sections as @rohitkrai03 completes the formik integration.

File: frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx (hunk 11 above)
Comment (debsmita1) [reply frontend/public/extend/devconsole/shared/components/advanced-options/Scaling.tsx]: Done

**Review by rohitkrai03 — APPROVED** [review]
LGTM

**Pull request merged**